# TensorFlow - Start

### Importing Tensorflow module in Python:

In [1]:
import tensorflow as tf
print("tensorflow version:" , tf.__version__)

tensorflow version: 2.13.0


#### The TensorFlow version is imported successfuly and the version is printed.

### Load a Dataset

In [3]:
# loading the mnist dataset
mnist = tf.keras.datasets.mnist

# preparing the mnist dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train/255.0, x_test/255.0

11490434/11490434 [==============================] - 3s 0us/step


### Building a machine learning #Model

In [10]:
# build a tf.keras.Sequantial Model:

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

### Predictions:

In [11]:
predictions = model(x_train[:1]).numpy()
predictions

array([[ 0.68727994,  1.2160454 ,  0.31223556,  0.00678589, -0.36322206,
         0.04953711, -0.46396035, -0.692292  , -0.46371078, -0.25100526]],
      dtype=float32)

#### The tf.nn.softmax fxn converts these logits -> to -> probabilities for each class:

In [12]:
tf.nn.softmax(predictions).numpy()

array([[0.16544622, 0.28073514, 0.11370438, 0.08377647, 0.05786682,
        0.08743567, 0.05232142, 0.04164054, 0.05233448, 0.06473882]],
      dtype=float32)

#### Defining a loss fxn for training using:
##### losses.SparseCategoricalCrossentropy

In [16]:
loss_fxn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

#### This untrained model gives probabilities close to random (1/10 for each class), so the initial loss should be close to:
###### tf.math.log(1/10)~=2.3

In [20]:
loss_fxn(y_train[:1], predictions).numpy()

2.436852

#### Before we start training, we need to:
###### 1. Configure the model
###### 2. Compile the model
##### using Keras model.compile

In [22]:
'''
set the optimizer class to 'adam'
set the loss to loss_fxn function defined earlier
specify a metric to be evaluated for the model by setting (metrics) parameter to (accuracy)
'''

model.compile(optimizer='adam',
              loss=loss_fxn,
              metrics=['accuracy'])

### Now we train and evaluate the model

###### We use model.fit fxn to adjust the model parameters and minimize the loss:

In [23]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 2s 960us/step - loss: 0.3012 - accuracy: 0.9127
Epoch 2/5
1875/1875 [==============================] - 2s 949us/step - loss: 0.1468 - accuracy: 0.9566
Epoch 3/5
1875/1875 [==============================] - 2s 950us/step - loss: 0.1118 - accuracy: 0.9661
Epoch 4/5
1875/1875 [==============================] - 2s 945us/step - loss: 0.0873 - accuracy: 0.9737
Epoch 5/5
1875/1875 [==============================] - 2s 961us/step - loss: 0.0758 - accuracy: 0.9764


#### Now, we use model.evaluate fxn to check the models performance on a 'validation set' OR 'test set'

In [24]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.0718 - accuracy: 0.9787 - 250ms/epoch - 799us/step


[0.07183584570884705, 0.9786999821662903]

#### The image classifier is trained at 0.9787 , i.e., with an accuracy of 97.87 ~= 98% .

#### Now, if we want the trained model to return a probability, we can wrap the trained model and attach softmax to it:

In [26]:
probability_model = tf.keras.Sequential([
    model,
    tf.keras.layers.Softmax()
])

##### Now we see the probability ->

In [27]:
probability_model(x_test[:5])

<tf.Tensor: shape=(5, 10), dtype=float32, numpy=
array([[3.30248251e-09, 9.69846692e-09, 2.47946555e-05, 9.17629368e-05,
        2.39208080e-11, 8.46787529e-08, 3.20074723e-13, 9.99877691e-01,
        1.18973531e-07, 5.46497267e-06],
       [5.85465045e-08, 1.48404331e-04, 9.99180019e-01, 6.68672263e-04,
        1.92725597e-15, 9.84729382e-08, 5.66641800e-09, 1.12571419e-11,
        2.71972272e-06, 1.65214129e-13],
       [2.13075623e-07, 9.98647034e-01, 2.32805105e-05, 6.27706095e-06,
        1.80248171e-05, 1.45405611e-05, 1.81658761e-05, 9.60533624e-04,
        3.10828065e-04, 1.19544495e-06],
       [9.99736607e-01, 1.74669079e-09, 1.10459354e-04, 1.50155097e-06,
        1.13541648e-06, 1.93886967e-06, 1.21613339e-04, 3.89079923e-06,
        5.57922306e-07, 2.21546179e-05],
       [2.01655780e-06, 1.70777170e-08, 2.35148286e-06, 7.76538300e-08,
        9.97404635e-01, 5.02808518e-07, 2.77581108e-07, 1.34214351e-05,
        5.81478844e-06, 2.57085636e-03]], dtype=float32)>

# Hence, we have successfully completed the training of an ML model using the keras api.